src/Microsoft.AspNetCore.Authentication/AuthenticationMiddleware.cs

```diff
@@ -2,90 +2,64 @@
         {
-            if (next == null)
-            {
-                throw new ArgumentNullException(nameof(next));
-            }
-
-            if (options == null)
-            {
-                throw new ArgumentNullException(nameof(options));
-            }
-
-            if (loggerFactory == null)
-            {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
-
-            if (encoder == null)
-            {
-                throw new ArgumentNullException(nameof(encoder));
-            }
-
-            Options = options.Value;
-            Logger = loggerFactory.CreateLogger(this.GetType().FullName);
-            UrlEncoder = encoder;
-
-            if (string.IsNullOrEmpty(Options.ClaimsIssuer))
-            {
-                // Default to something reasonable
-                Options.ClaimsIssuer = Options.AuthenticationScheme;
-            }
-
-            _next = next;
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            Schemes = schemes ?? throw new ArgumentNullException(nameof(schemes));
         }
 
-        public string AuthenticationScheme { get; set; }
-
-        public TOptions Options { get; set; }
-
-        public ILogger Logger { get; set; }
-
-        public UrlEncoder UrlEncoder { get; set; }
+        public IAuthenticationSchemeProvider Schemes { get; set; }
 
         public async Task Invoke(HttpContext context)
         {
-            var handler = CreateHandler();
-            await handler.InitializeAsync(Options, context, Logger, UrlEncoder);
+            var oldFeature = context.Features.Get<IAuthenticationFeature>();
             try
             {
-                if (!await handler.HandleRequestAsync())
+                context.Features.Set<IAuthenticationFeature>(new AuthenticationFeature
                 {
-                    await _next(context);
-                }
-            }
-            finally
-            {
-                try
+                    OriginalPath = context.Request.Path,
+                    OriginalPathBase = context.Request.PathBase
+                });
+
+                var defaultAuthenticate = await Schemes.GetDefaultAuthenticateSchemeAsync();
+                if (defaultAuthenticate != null)
                 {
-                    await handler.TeardownAsync();
+                    var result = await context.AuthenticateAsync(defaultAuthenticate.Name);
+                    if (result?.Principal != null)
+                    {
+                        context.User = result.Principal;
+                    }
                 }
-                catch (Exception)
+
+                // REVIEW: alternatively could depend on a routing middleware to do this
+
+                // Give any IAuthenticationRequestHandler schemes a chance to handle the request
+                var handlers = context.RequestServices.GetRequiredService<IAuthenticationHandlerProvider>();
```

Comment:
Move HandleRequestAsync processing above defaultAuthenticate. HandleRequestAsync may short circuit the request.